plugin/sublime-text/c_api_subl3.py

```diff
@@ -0,0 +1,185 @@
+
+
+class SeverityEnumeration:
+    ERROR = 1
+    WARNING = 2
+
+
+class DiagnosticStructure(ctypes.Structure):
+    """Diagnostic layer used to communicate with the C++ code."""
+
+    # struct qljs_sublime_text_3_diagnostic {
+    #   const char* message;
+    #   const char* code;
+    #   qljs_severity severity;
+    #   int begin_offset;
+    #   int end_offset;
+    # };
+    _fields_ = [
+        ("message", ctypes.c_char_p),
+        ("code", ctypes.c_char_p),
+        ("severity", ctypes.c_int),
+        ("begin_offset", ctypes.c_int),
+        ("end_offset", ctypes.c_int),
+    ]
+
+
+DiagnosticStructurePointer = ctypes.POINTER(DiagnosticStructure)
+
+
+class ParserStructure(ctypes.Structure):
+    """Parser layer used to communicate with the C++ code."""
+
+    _fields_ = []
+
+
+ParserStructurePointer = ctypes.POINTER(ParserStructure)
+
+
+def get_script_path_directory():
+    return os.path.dirname(os.path.realpath(__file__))
+
+
+def get_parent_directory(path, level=1):
+    return os.path.normpath(os.path.join(path, *([".."] * level)))
+
+
+def load_library():
+    # If the plugin is inside a package you will need to unzip the
+    # shared library from inside the package before continuing.
+    #
+    # From:
+    # |-- quick-lint-js-v020-linux_x86_64.sublime-package
+    #
+    # To:
+    # |-- quick-lint-js-v020-linux_x86_64.sublime-package
+    # |-- quick-lint-js
+    #     |-- libquick-lint-js-lib.so
+    #
+    script_path_directory = get_script_path_directory()
+    if script_path_directory.endswith(".sublime-package"):
+        with zipfile.ZipFile(script_path_directory, mode="r") as package_file:
+            lib_path_directory = get_parent_directory(script_path_directory)
+            lib_path_directory += "/quick-lint-js"
+            package_file.extract(
+                "libquick-lint-js-lib.so",
+                path=lib_path_directory,
+            )
+    else:
+        lib_path_directory = script_path_directory
+
+    # Load the shared library.
+    lib_path = lib_path_directory + "/libquick-lint-js-lib.so"
+    lib = ctypes.CDLL(lib_path)
+    return lib
+
+
+def set_argtypes_and_restype(func, argtypes, restype):
+    func.argtypes = argtypes
+    func.restype = restype
+
+
+def create_library():
+    lib = load_library()
+    set_argtypes_and_restype(
+        lib.qljs_sublime_text_3_create_parser,
+        argtypes=[],
+        restype=ParserStructurePointer,
+    )
+    set_argtypes_and_restype(
+        lib.qljs_sublime_text_3_destroy_parser,
+        argtypes=[ParserStructurePointer],
+        restype=None,
+    )
+    set_argtypes_and_restype(
+        lib.qljs_sublime_text_3_set_text,
+        argtypes=[ParserStructurePointer, ctypes.c_void_p, ctypes.c_size_t],
+        restype=None,
+    )
+    set_argtypes_and_restype(
+        lib.qljs_sublime_text_3_lint,
+        argtypes=[ParserStructurePointer],
+        restype=DiagnosticStructurePointer,
+    )
+    return lib
+
+
+LIB = create_library()
```

Comment:
Not yet. However, it is relevant that the error message, be helpful or not, will depend more on **Python** and the user's **operating system** than on me. That's because the error message that Sublime Text would present to the user through a pop-up would be the exception message that the [ctypes] function threw when loading the library. The only two possible exceptions when loading the shared library are [OSError] and [WindowsError]. These exceptions receive their messages from the **operating system**.

[ctypes]: https://docs.python.org/3/library/ctypes.html#loading-dynamic-link-libraries
[OSError]: https://docs.python.org/3/library/exceptions.html#OSError
[WindowsError]: https://docs.python.org/3/library/exceptions.html#WindowsError

Reply:
Now yes, users receive a pop-up with the error message:

55485e65e99a44ddb5de319a0e48533b5bdbf410
30d7dd87803aff75d3306f82f41612a721ecf419